Fix locations

jioter opened the pull request:
Fix locations

jioter committed:
Remove unnecessary validation in LocationBaseVO

jioter committed:
Post Location Endpoint
- check by all fields if entity exists in DB if yes - return it if no - create new

jioter committed:
Fixed endpoint findAllCitiesByCountryCode

return Set of Cities without empty one and null's

jioter committed:
Add sorting to findAllCitiesByCountryCode Location Service

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/repository/LocationRepository.java:
```diff
@@ -13,4 +13,6 @@
 
     List<Location> findAllByCity(String city);
+
+    Location findByCountry_CountryCodeAndCityAndAddressLine(String countryCode, String city, String addressLine);
```
For now this line can potentially return several Locations with the same countryCode - city - addressLine.
Let's consider adding unique composite index for (countryCode - city - addressLine) to the location

jioter replied:
Added, but commented

jioter replied:
In future we need somehow, as u said find duplicates and remove them

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/service/LocationService.java:
```diff
@@ -49,14 +54,20 @@ public Location findEntityById(Long id) {
-        locations.forEach(entity -> response.add(cityMapper.toVO(entity)));
-        return response;
+        List<CityResponseVO> allItems = new ArrayList<>();
+        locations.forEach(entity -> allItems.add(cityMapper.toVO(entity)));
```
It's better to filter all locations before mapping them with cityMapper. We can refactor it. Let's discuss on our daily call

jioter replied:
Refactored to return list of strings

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/service/LocationService.java:
```diff
@@ -49,14 +54,20 @@ public Location findEntityById(Long id) {
     @Transactional(propagation = Propagation.REQUIRED, readOnly = true)
     public List<CityResponseVO> findAllCitiesByCountryCode(String countryCode) {
         List<Location> locations = locationRepository.findAllByCountry_CountryCode(countryCode);
```
We are retrieving too much data here (which is not needed). In the nearest future we can refactor it to retrieve only what we need (custom projections?)

jioter replied:
Added protection, but I'm not sure if it's done right.

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/service/LocationService.java:
```diff
@@ -49,14 +54,20 @@ public Location findEntityById(Long id) {
+        List<CityResponseVO> allItems = new ArrayList<>();
+        locations.forEach(entity -> allItems.add(cityMapper.toVO(entity)));
+        return allItems.stream()
+                .filter(el -> el.getCity() != null && !el.getCity().equals("")).distinct()
```
it's better use StringUtils.isNotBlank( ... ) from the commons lib

jioter replied:
Used, thanks

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/service/LocationService.java:
```diff
@@ -49,14 +54,20 @@ public Location findEntityById(Long id) {
+                request.getCountryCode(),
+                request.getCity(),
+                request.getAddressLine());
+        return location == null ? locationMapper.toVO(locationRepository.save(locationMapper.toEntity(request))) : locationMapper.toVO(location);
```
It is more readable to write these methods in separate lines

jioter replied:
I have used new line, coz if declarate new 2 variables and use them in ternary operator - then we get 500 when entity exits on third time, dunno yet, why. Returned to initial state.

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/service/LocationService.java:
```diff
@@ -9,6 +9,8 @@
 import com.exadel.discountwebapp.location.vo.city.CityResponseVO;
 import com.exadel.discountwebapp.location.vo.location.LocationRequestVO;
 import com.exadel.discountwebapp.location.vo.location.LocationResponseVO;
+import com.google.common.collect.Lists;
```
I see some Sonar issues in this class

jioter replied:
Fixed - removed

jioter committed:
Merge branch 'master' of https://github.com/dmitriychernov7/sandbox-team3-be into fix-locations

jioter committed:
Add unique constraint to Location (country code, city, address line)

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/service/LocationService.java:
```diff
@@ -49,14 +54,20 @@ public Location findEntityById(Long id) {
     }
 
     @Transactional(propagation = Propagation.REQUIRED)
     public LocationResponseVO create(LocationRequestVO request) {
```
let's rename to createIfNotExist

jioter replied:
Renamed

jioter committed:
Add projection for City

• Removed unused code

• getAllCitiesByCountry - now return List of Strings

• Added projection to Location -> City

• refactored stream in findAllCitiesByCountryCode

jioter committed:
Merge branch 'master' of https://github.com/dmitriychernov7/sandbox-team3-be into fix-locations

 Conflicts:
	src/main/resources/db/changelog/db.changelog-master.yml

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/controller/LocationController.java:
```diff
@@ -44,15 +43,15 @@ public LocationResponseVO getLocationById(@PathVariable Long id) {
 
     // TODO: tests for this controller
     @GetMapping("/cities")
-    public List<CityResponseVO> getAllCitiesByCountry(
```
it seems like will break the UI after this change

jioter replied:
We just have called concerning it and they say it's okay, they will modify on front to process List of Strings

ahlukhouski2 commented on src/main/java/com/exadel/discountwebapp/location/service/LocationService.java:
```diff
@@ -3,29 +3,28 @@
 import com.exadel.discountwebapp.location.vo.location.LocationRequestVO;
 import com.exadel.discountwebapp.location.vo.location.LocationResponseVO;
 import lombok.RequiredArgsConstructor;
+import org.codehaus.plexus.util.StringUtils;
```
wrong library, let's use org.apache.commons.lang3.StringUtils;

jioter replied:
Changed to good one)

jioter committed:
Small fixes

import
